feat: Add product comparison feature

Sunhao86 opened the pull request:
### What
This PR adds a product comparison feature to allow users to compare multiple food products side by side

### Screenshot
![截图 2025-03-27 20-16-48](https://github.com/user-attachments/assets/b7ff45a4-33a1-4d07-b3ec-1c61115649a2)
![截图 2025-03-27 20-16-53](https://github.com/user-attachments/assets/8bd05c18-33e0-4a98-b674-6a2b9c945713)


### Features
- Users can add products to a comparison list from product detail pages
- Users can view all selected products in a comparison view
- Users can remove products or clear the entire list

Sunhao86 commented:
I've implemented this product comparison feature to help users compare food products side by side. I genuinely welcome any feedback or suggestions for improvement. Thank you for reviewing my PR 🙏

teolemon commented:
- Can you add the screenshot of the compare system as well ? @Sunhao86 
- One interesting thing would be infinite side by side comparison, but the product battle mode is cool already.

VaiTon commented on src/lib/stores/compareStore.ts:
```diff
@@ -0,0 +1,51 @@
+
+	if (browser) {
+		subscribe((products) => {
+			localStorage.setItem('compareProducts', JSON.stringify(products));
```
Instead of this please use `stored` with `svelte-local-storage-store`

Sunhao86 commented:
Hi @VaiTon ,I've updated the compareStore to use `svelte-local-storage-store` as requested. Please kindly review again 🙏

Sunhao86 commented:
Hi @teolemon ,Thank you for the feedback!  As for the screenshot, here’s how the current comparison system looks:

Currently, the comparison supports up to 4 products side by side. I’ve limited it to 4 because adding more makes the layout messy, especially on smaller screens. But I'm happy to explore infinite scroll-style layouts in the future!

Let me know if you'd like me to tweak anything further 🙏
![image](https://github.com/user-attachments/assets/0a455e76-4925-41e9-a504-200289f2d5dc)

VaiTon commented on src/lib/ui/CompareFloatingButton.svelte:
```diff
@@ -0,0 +1,20 @@
+	let count = $state(0);
+
+	$effect(() => {
+		count = $compareStore.length;
```
This should be a $derived

VaiTon commented on src/routes/compare/+page.svelte:
```diff
@@ -0,0 +1,114 @@
+		<div class={isMobile ? 'overflow-x-auto pb-4' : ''}>
+			<div
+				class={isMobile ? 'w-max' : ''}
+				style={isMobile
+					? ''
+					: `display: grid; grid-template-columns: repeat(${$compareStore.length}, 1fr); gap: 1rem;`}
```
Please use TailwindCSS instead of custom styles. Also, is the isMobile var really necessary? Can't we do that in CSS?

VaiTon commented on src/routes/compare/+page.ts:
```diff
@@ -0,0 +1,8 @@
+
+export const load: PageLoad = async () => {
+	return {
+		title: 'Compare Products',
+		description: 'Compare nutritional information of food products'
```
What is the utility of this file? Couldn't we just embed the text inside the .svelte file?

teolemon commented:
What I'd like to see as screenshot is how you add a product to the comparison @Sunhao86

Sunhao86 commented:
hi @teolemon , here are the screenshots showing how a product is added to the comparison:

From the product detail page, I click on "Add to Compare"

Then, I go to the Compare Products page to see all selected products displayed side by side.
![截图 2025-04-04 16-07-29](https://github.com/user-attachments/assets/f331861b-d47e-4039-8ec4-5cdf6bfce415)
![截图 2025-04-04 16-09-15](https://github.com/user-attachments/assets/c3d23f85-a7c6-4550-b6b5-8c230d1834c1)

Sunhao86 commented:
Hi @VaiTon , I’ve implemented all your suggestions:

replaced $state with $derived

removed isMobile and switched to TailwindCSS

deleted the unused +page.ts

Everything works well locally. Please feel free to share more feedback — I’ll be happy to adjust anything needed. Thanks a lot!

VaiTon commented:
@github-throwaway @teolemon UX feedback?

github-throwaway commented:
Nice idea!

- It looks like the X symbol does not play nice with large images. I would maybe move that and make it more prominent
- I'm unsure about the rectangle images and the rounded bounding boxes. I feel that does not match the style we have on the landing page. But I also don't have a concrete solution to this. I just think it stands out.
- the drop shadow is missing if we wanna stay consistent with the landing page. 
- unrelated but we really need to fix the margins in the navbar. They are all over the place... 
- it terms of comparability it's a good start. But I would expect to see nutrition facts and ingredients side by side. The current state is a bit superficial (a great start though, well done!)
- what happens when I click clear all?
- what does this look like on mobile? (Product pages didn't load for me on iOS chrome)

cc: @abrahamvg

abrahamvg commented:
Suggestions:
- The container should have a fixed width otherwise it will expand and not look good when there are less than 4 items.
- The X can be wrapped inside a circle container with the bg being red
- `Hot take` : instead of the images for nutritional scores, can be add it as a value ? or we can stack it vertically to increase readability

github-throwaway commented:
Also what happens when I click the compare button in the lower right corner?

Bhavishya2601 commented:
we already have 4 buttons there, adding a 5th btn there (Add to compare), I think so would affect the UI

In this product if you see, It's already quite suffocated
![image](https://github.com/user-attachments/assets/f72e75b2-ab3b-4b40-9d84-70b716b4b3ed)


@Sunhao86 can you pls provide a screenshot of the product page where the btn (Add to compare) will be mounted after pulling all the new commits

swastikCommits commented:
We can replace the text inside these buttons by replacing it with the following logos:
Add to calculator: https://icon-sets.iconify.design/ic/?icon-filter=calc&keyword=p
Add to compare: https://icon-sets.iconify.design/ic/?icon-filter=compa&keyword=p
Edit: https://icon-sets.iconify.design/ic/?icon-filter=edit&keyword=p
Share: Just by removing the text inside the button

These buttons on hover must specify "Add to calculator". "Add to compare", "Edit", "Share"
wby @github-throwaway and @VaiTon ?

github-throwaway commented:
@Sunhao86 Please update your feature branch to include the latest changes of the main branch. 

On mobile we should use icons. And on desktop text + icons.

Sunhao86 commented:
> @Sunhao86 Please update your feature branch to include the latest changes of the main branch.
> 
> On mobile we should use icons. And on desktop text + icons.

Hi @github-throwaway , I've updated the feature branch with the latest changes from `main` and resolved the merge conflicts, as requested.

Sunhao86 commented:
Hi @VaiTon  @github-throwaway  ,  Thanks for the great suggestions!
We've updated the CompareFloatingButton to avoid overlapping with the calculator button on mobile view.
It now appears on the left bottom corner, while the calculator stays on the right. I'm really thank for your UI/UX suggestions!
![Screenshot 2025-04-13 170115](https://github.com/user-attachments/assets/676d39c3-e915-4e25-baca-dd65d358773a)
![Screenshot 2025-04-13 170100](https://github.com/user-attachments/assets/5168e94a-627b-46de-b0c1-be7d83183580)

Bhavishya2601 commented:
@Sunhao86 Why do we even need a floating button for compare? can't we place it next to folksonomy as it is redirecting us to compare page and is not a pop up like nutrition calculator

and pls fix this
![image](https://github.com/user-attachments/assets/d640c966-d07c-4cbb-a911-a8054676fc0d)


@VaiTon @github-throwaway don't you think compare feature and nutrition calculator are quite same in some perspectives
![image](https://github.com/user-attachments/assets/c6a78f6f-dc73-4cf2-8207-17c5ec53b058)

Sunhao86 commented:
@Bhavishya2601 Thanks for the feedback! Thank you very much for your advice!

I understand the concern about having multiple floating buttons, especially on mobile where screen space is limited.

As an alternative, I was wondering if we could move the compare functionality to a more dedicated place — for example, a separate page 
This would avoid floating buttons altogether while still preserving quick access to comparisons.

What do you think about this direction? @github-throwaway

github-throwaway commented:
It might make sense to combine these things. They are indeed quite comparable.

 The dedicated page does not make sense to me. I think most features should be available from the product page directly.

VaiTon commented:
> @VaiTon @github-throwaway don't you think compare feature and nutrition calculator are quite same in some perspectives

Not necessarily, but if we find a way to integrate them I'm for it

github-throwaway commented:
please provide a mockup of your alternative compare page. You are allowed to dislike something but that’s not constructive criticism. Instead address concrete pain points. 

The nutritional info is not easily comparable. The info should be on the same row. Maybe some sort of table view would make more sense here?

Sunhao86 commented:
@github-throwaway  @VaiTon @Bhavishya2601  Thank you all so much for your valuable comments and suggestions recently!
@github-throwaway Thank you for further elaborating on the idea of ​​a dedicated compare page and your preference for integrating the functionality directly into the product page. However, in order to better realize your vision, I would like to ask:
1. Regarding "providing functionality directly from the product page", what specific interaction method do you prefer? For example, is it to pop up a dialog box (modal) to display the comparison information after clicking a button on the product page? Or dynamically embed a comparison area in the current page? Or are there other ways or reference examples that you think are better for the experience that you can share?
2. You mentioned using a table view (rows are nutrients, columns are products) to compare nutritional information. This idea sounds great and is indeed more conducive to direct comparison. Is this table view also used in your envisioned product page integration solution?
@VaiTon  @Bhavishya2601 Thank you for your thoughts on integrating the comparison and calculator functions and your feedback on the current comparison page UI. If we determine a new design for product page integration, we can discuss how to better combine these ideas.

Thanks again for everyone's help!

teolemon commented:
@VaiTon on mobile yes
<img width="1500" height="3248" alt="Lists - Light - En" src="https://github.com/user-attachments/assets/e8e6eb56-a45c-4736-a8ab-7739e091f8a2" />
<img width="1500" height="3248" alt="Side by side Compare - Light - EN-1" src="https://github.com/user-attachments/assets/b6cda293-e1cc-4093-a5a9-3b0bc72741b2" />
<img width="1500" height="3248" alt="Side by side Compare - Light - EN" src="https://github.com/user-attachments/assets/fed5dfb8-f2e4-400b-a857-5ba08e264837" />
